stages/org.osbuild.tmpfilesd

```diff
@@ -0,0 +1,121 @@
+
+
+SCHEMA = r"""
+"definitions": {
+  "configuration": {
+    "type": "object",
+    "additionalProperties": false,
+    "required": ["type", "path"],
+    "description": "tmpfiles.d configuration directive representing one line in the configuration.",
+    "properties": {
+      "type": {
+        "type": "string",
+        "description": "The file system path type.",
+        "pattern": "^([fwpLcbaA]\\+?|[dDevqQCxXrRzZtThH]){1}((!?-?)|(-?!?)){0,1}$"
+      },
+      "path": {
+        "type": "string",
+        "description": "Absolute file system path."
+      },
+      "mode": {
+        "type": "string",
+        "description": "The file access mode when creating the file or directory.",
+        "pattern": "^~?[0-7]{4}$"
+      },
+      "user": {
+        "type": "string",
+        "description": "The user to use for the file or directory."
+      },
+      "group": {
+        "type": "string",
+        "description": "The group to use for the file or directory."
+      },
+      "age": {
+        "type": "string",
+        "description": "Date field used to decide what files to delete when cleaning."
+      },
+      "argument": {
+        "type": "string",
+        "description": "Argument with its meaning being specific to the path type."
+      }
+    }
+  }
+},
+"additionalProperties": false,
+"required": ["filename", "config"],
+"properties": {
+  "filename": {
+    "type": "string",
+    "description": "Name of the tmpfiles.d configuration file to create.",
+    "pattern": "^[\\w.-]{1,250}\\.conf$"
+  },
+  "config": {
+    "additionalProperties": false,
+    "type": "array",
+    "description": "List of configuration directives written into the configuration file.",
+    "minItems": 1,
+    "items": {
+      "$ref": "#/definitions/configuration"
+    }
+  }
+}
+"""
+
+
+def main(tree, options):
+    filename = options["filename"]
+    cfg = options["config"]
+
+    tmpfilesd_config_dir = f"{tree}/usr/lib/tmpfiles.d"
+
+    cfg_lines = []
+    for cfg_item in cfg:
+        cfg_type = cfg_item["type"]
+        cfg_path = cfg_item["path"]
+
+        cfg_line = f"{cfg_type} {cfg_path}"
+
+        optional_properties = ["mode", "user", "group", "age", "argument"]
+        cfg_line_optional_part = ""
+        for optional_property in reversed(optional_properties):
+            cfg_property_value = cfg_item.get(optional_property)
+            if cfg_property_value:
+                if cfg_line_optional_part:
+                    cfg_line_optional_part = " ".join([cfg_property_value, cfg_line_optional_part])
+                else:
+                    cfg_line_optional_part = cfg_property_value
+            elif cfg_line_optional_part:
+                # if there were already some optional properties provided, then
+                # we must use "-" for any not provided optional values preceding
+                # them on the configuration line.
+                cfg_line_optional_part = " ".join(["-", cfg_line_optional_part])
```

Comment:
I found this a bit hard to follow at first.  Would it work the same if it was simplified by simply reading:
```python
cfg_property_value = cfg_item.get(optional_property, "-")
```
and then joining all property values unconditionally with spaces?

Reply:
I admit that this is a bit hard to follow, but I didn't find a more elegant way to do it.

The thing is that while I could do what you are suggesting, the outcome would be AFAIK functionally the same, but would look different to what people may expect and to what the man page shows as examples.

The structure of a line in the configuration is as follows:
```
<type>  <path>  <mode>  <user>  <group>  <age>  <argument>
```

Technically only `type` and `path` must be provided on the line. The rest of the values can be omitted or not specified using `-`. Now, the practice seems to be to omit all values if not provided, unless a previous "optional" value (taken from the right to left) has been provided. Since the meaning of values is determined by their position on the line (after the line is split by whitespaces), using `-` for omitted values, which are "in between" provided values is necessary.

So in the current implementation, the following stage:
```json
{
  "name": "org.osbuild.tmpfilesd",
  "options": {
    "filename": "example.conf",
    "config": [
      {
        "type": "w-",
        "path": "/proc/sys/vm/swappiness",
        "argument": "10"
      },
      {
        "type": "d",
        "path": "/tmp/.X11-unix"
      },
      {
        "type": "d",
        "path": "/run/uscreens",
        "mode": "0755",
        "group": "screen"
      }
    ]
  }
}
```

results in an `example.conf` file with the following content:
```
w- /proc/sys/vm/swappiness - - - - 10
d /tmp/.X11-unix
d /run/uscreens 0755 - screen
```

while with your suggestion, it would be:
```
w- /proc/sys/vm/swappiness - - - - 10
d /tmp/.X11-unix - - - - -
d /run/uscreens 0755 - screen - -
```

The problem here may be that the `tmpfiles.d(5)` man page does not specify that `-` can be used for the `<argument>` in the meaning of an omitted value.

I'm happy to rework the code in some way to make it easier to understand if you have any suggestions. The issue is that  simply using `-` for all omitted values it not feasible (mainly due to the `<argument>` value). So at least the `<argument>` value would need to be handled separately from the rest. Alternatively I can add additional comment to explain the intention a bit better.

Reply:
I see.  I think it's okay to merge as is then and keep the resulting files cleaner and closer to how they're conventionally written.